src/language-markdown/whitespace.js

```diff
@@ -0,0 +1,265 @@
+      }
+    }
+  }
+  // Injects a property to cache the result.
+  sentenceNode.usesCJSpaces =
+    cjNonCJKSpacingStatistics[" "] > cjNonCJKSpacingStatistics[""];
+  return sentenceNode.usesCJSpaces;
+}
+
+/**
+ * @typedef {import("./utils.js").TextNode} TextNode
+ * @typedef {import("./utils.js").WhitespaceValue} WhitespaceValue
+ * @typedef {{next?: TextNode | undefined | null, previous?: TextNode | undefined | null}} AdjacentNodes
+ * @typedef {import("./utils.js").WordKind} WordKind
+ * @typedef {import("../common/ast-path.js").default} AstPath
+ */
+
+/**
+ * Checks if given node can be converted to Space
+ *
+ * For example, if you would like to squash the multi-line string `"You might want\nto use Prettier."` into a single line,
+ * you would replace `"\n"` with `" "`. (`"You might want to use Prettier."`)
+ *
+ * However, you should note that Chinese and Japanese does not use U+0020 Space to divide words, so U+000A End of Line must not be replaced with it.
+ * Behavior in other languages (e.g. Thai) will not be changed because there are too much things to consider. (PR welcome)
+ *
+ * @param {AstPath} path path of given node
+ * @param {WhitespaceValue} value value of given node (typically `" "` or `"\n"`)
+ * @param {AdjacentNodes | undefined} adjacentNodes adjacent sibling nodes of given node
+ * @returns {boolean} `true` if given node can be converted to space, `false` if not (i.e. newline or empty character)
+ */
+function canBeConvertedToSpace(path, value, adjacentNodes) {
+  // "\n" or " ", of course " " always can be converted to Space
+  if (value !== "\n") {
+    return true;
+  }
+  // no adjacent nodes
+  if (!adjacentNodes) {
+    return true;
+  }
+  // e.g. " \nletter"
+  if (!adjacentNodes.previous || !adjacentNodes.next) {
+    return true;
+  }
+  const previousKind = adjacentNodes.previous.kind;
+  const nextKind = adjacentNodes.next.kind;
+  // "\n" between not western or Korean (han, kana, CJK punctuations) characters always can converted to Space
+  // Korean hangul simulates latin words; see #6516 (https://github.com/prettier/prettier/issues/6516)
+  if (
+    isWesternOrKoreanLetter(previousKind) &&
+    isWesternOrKoreanLetter(nextKind)
+  ) {
+    return true;
+  }
+  // han & hangul: same way preferred
+  if (
+    (previousKind === KIND_K_LETTER && nextKind === KIND_CJ_LETTER) ||
+    (nextKind === KIND_K_LETTER && previousKind === KIND_CJ_LETTER)
+  ) {
+    return true;
+  }
+  // Do not convert it to Space when:
+  if (
+    // Shall not be converted to Space around CJK punctuation
+    previousKind === KIND_CJK_PUNCTUATION ||
+    nextKind === KIND_CJK_PUNCTUATION ||
+    // "\n" between CJ always SHALL NOT be converted to space
+    // "\n" between Korean and CJ is better not to be converted to space
+    (isCJK(previousKind) && isCJK(nextKind))
+  ) {
+    return false;
+  }
+  const previousLastChar = adjacentNodes.previous.value?.at(-1);
+  const nextFirstChar = adjacentNodes.next.value?.[0];
```

Comment:
Is it really possible that `value` can be undefined?

Reply:
Assumed a sequence of whitespace nodes (previous or next is whitespace).

Reply:
I'm asking to understand whether optional chaining is really needed here.

Reply:
@thorn0 I will remove optional chaining if you say both of previous and next are always word nodes in this line.

Reply:
I checked the logic in `splitText`. It doesn't create consecutive whitespace nodes. However, to stay safe in the future, please enforce this constrain by adding the following check to `splitText` before returning `result`:

```js
  if (process.env.NODE_ENV !== "production") {
    for (let i = 1; i < nodes.length; i++) {
      if (
        nodes[i].type === "whitespace" &&
        nodes[i - 1].type === "whitespace"
      ) {
        throw new Error(
          "splitText should not create consecutive whitespace nodes"
        );
      }
    }
  }
```

Reply:
@thorn0 I also checked it and confirmed that a whitespace node is not adjacent to another one. I fixed this.